core/common/src/main/java/alluxio/collections/IndexDefinition.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+/**
+ * An class representing the definition of an index for this {@link IndexedSet}, each index for
+ * this set must implement the interface to define how to get the value of the field chosen as
+ * the index. Users must use the index name defined in this class as the
+ * parameter in all methods of {@link IndexedSet} to represent the same index.
+ *
+ * @param <T> type of objects in this {@link IndexedSet}
+ */
+public class IndexDefinition<T> {
+
+  /**
+   * An interface abstracting the value of the field that serves as index.
+   *
+   * @param <T> type of objects in this {@link Abstracter}
+   */
+  public interface Abstracter<T> {
+    /**
+     * Gets the value of the field that serves as index.
+     *
+     * @param o the instance to get the field value from
+     * @return the field value, which is just an Object
+     */
+    Object getFieldValue(T o);
+  }
+
+  /** The index name, representing the index in methods of {@link IndexedSet}. */
+  private final String mName;
```

Comment:
Why not identify index definitions by reference like we did with `FieldIndex` before?